### PR TITLE
900MHz Maximum Power TX/RX

### PR DESCRIPTION
**Pull request by Gorilla-Link**
Maximum Power for ELRS 900 MHz Hardware (TX and RX)

Hardware included:
Generic 900 MHz
FrSky R9

Settings are changed with the lua script. (version 3)

### CODE REVIEW

**Comment by SunjunKim** [comment]
It's an abuse of the ELRS Github repo. Don't do this.

**Comment by Gorilla-Link** [comment]
> It's an abuse of the ELRS Github repo. Don't do this.

Please explain??

**Comment by Gorilla-Link** [comment]
It can be merged if necessary, why not allow for discussion at least?

**Comment by SunjunKim** [comment]
PR meant to be you want to add some feature to the mainline code, and make a contribution at the end.

This PR's purpose is only to make your target (somehow) available on the configurator, which we consider an abuse.

If you want the target on the configurator officially, please join our discord, and contact one of the devs for the proper collaboration.

**Comment by SunjunKim** [comment]
> It can be merged if necessary, why not allow for discussion at least?

We can discuss, if you have any topics other than "only needs to exist as pull request for configurator"

**Comment by SunjunKim** [comment]
As far as I understand, the main contribution of this PR is to unlock 2W power from R9 TX?
I'm wondering, have you measured the actual power generated, and is it stable over time?

**Comment by JyeSmith** [comment]
These settings have a restriction of 1% duty cycle, where ELRS is closer to 80%.

Do you have any measurements of the increased power output?

**Comment by Gorilla-Link** [comment]
We actually have been on the discord for well over a year, and indeed have been in discussion with one of the original developers, Wez, who suggested we initiate this pull request, if not for open discussion sake.
Our understanding is that is how pull requests are supposed to be used, tested and verified BEFORE being merged,
I'm sorry to say, Sunjun, you have misunderstood and abused your power to close something without even attempting a discussion.
Hopefully Sunjun will be more open for community based projects in the future.

Thank you for the constructive questions, Jye, the duty cycle is higher than the Sx127X manual's threshold suggests, however we have tested in the field with consistent results. While the power output is typically less than the advertised maximum, it is still a significant increase from the non-boosted settings.

We hope this PR can remain open for others to test and develop.

**Comment by Gorilla-Link** [comment]
Thank you, Sunjun, for taking more time and consideration for this pull request. The wording of the original comment was perhaps written hastily, and as a result was unclear. Although, to be completely sure we avoid any bad impressions, is there any set of rules or guidelines specifically for ELRS pull requests, beyond the general ones set forth by github?

**Comment by JyeSmith** [comment]
> Sunjun, you have misunderstood and abused your power to close something without even attempting a discussion. Hopefully Sunjun will be more open for community based projects in the future.

Sunjuns feels are shared by a few Devs, and he beat me to it.

Users have come to see the Configurator as a place where tested hardware is listed.  Creating a PR with the title "Gorilla-Link" is a way to get you targets easily identified in the Configurator and bypass any checks ELRS have in place.  Coupled with the fact we have looked at higher output in the past and 100mW/2W were not achieved.  Happy to revisit this but it needs to be confirmed the output is correct AND hardware isn't being damaged.

The initial feel of this PR was about you the Vendor, and not for the communities benefit.

> Thank you for the constructive questions, Jye, the duty cycle is higher than the Sx127X manual's threshold suggests, however we have tested in the field with consistent results. While the power output is typically less than the advertised maximum, it is still a significant increase from the non-boosted settings.

An increase above 1W, and not achieve 2W, isn't a reason to add a 2W setting.  This is the quickest way to upset users since they will test the output and feel cheated due to advertised specs not matching reality.  Plenty of hard can go above their highest output, but they do not reach the next 3dBm setting.

Can you provide some screen shots of the receiver/transmitters being tested with the high output.  Not only at boot put after an extended time of use.

File: src/include/target/Frsky_TX_R9M.h [hunk 1]
```diff
@@ -42,9 +42,9 @@
...
 #define MinPower                        PWR_10mW
-#define HighPower                       PWR_250mW
-#define MaxPower                        PWR_1000mW
+#define HighPower                       PWR_2000mW
+#define MaxPower                        PWR_2000mW
 #if !defined(POWER_OUTPUT_VALUES)
-    #define POWER_OUTPUT_VALUES         {720,875,1000,1140,1390,1730,2100,2600}
-    #define POWER_OUTPUT_VALUES_868     {650,860,1000,1160,1420,1730,2100,2600}
+    #define POWER_OUTPUT_VALUES         {720,875,1000,1140,1390,1730,2300,3300}
+    #define POWER_OUTPUT_VALUES_868     {650,860,1000,1160,1420,1730,2300,3300}
```
Comment (JyeSmith) [inline src/include/target/Frsky_TX_R9M.h:49]: This looks to be driving the V_APC beyond its max of 2.8V and not much is achieved going beyond the typical.

image

image

File: src/include/target/Frsky_TX_R9M_LITE_PRO.h [hunk 2]
```diff
@@ -36,7 +36,7 @@
 
 // Output Power
 #define MinPower                    PWR_10mW
-#define HighPower                   PWR_250mW
-#define MaxPower                    PWR_1000mW
+#define HighPower                   PWR_2000mW
+#define MaxPower                    PWR_2000mW
 #define POWER_OUTPUT_ANALOG
-#define POWER_OUTPUT_VALUES         {600,770,950,1150,1480,2000,3500}
+#define POWER_OUTPUT_VALUES         {600,770,950,1150,1480,2000,3500,4095}
```
Comment (JyeSmith) [inline src/include/target/Frsky_TX_R9M_LITE_PRO.h:42]: Same here, V_APC is above the max of 2.8V.

File: src/lib/SX127xDriver/SX127x.cpp [hunk 3]
```diff
@@ -193,11 +206,28 @@ void SX127xDriver::SetOutputPower(uint8_t Power)
...
-  pwrCurrent = pwrPending;
-  pwrPending = PWRPENDING_NONE;
-  hal.writeRegister(SX127X_REG_PA_CONFIG, pwrCurrent);
+  }
+  if (SX127X_PA_20DBM == true)
+  {//20dbm
+    pwrCurrent = pwrPending;
+    pwrPending = PWRPENDING_NONE;
+    //20dbm
+    hal.writeRegister(SX127X_REG_PA_DAC, 0x87);  //turn on 3rd amplifier, see 5.4.3
```
Comment (JyeSmith) [inline src/lib/SX127xDriver/SX127x.cpp:217]: This register has a restriction of 1% duty cycle which ELRS goes well beyond.

image

**Comment by Gorilla-Link** [comment]
Thank you for excellent discussion points, Jye! Like we said before, the extensive field testing has been done on our end with great results, so we are hoping to bring that to everyone for FREE with this pull request.
It seems like the other vendors have left the developers in a state of distrust, and we are sorry to hear that. Like you mentioned, the bench testing metrics can discourage some new users, but of course we are only working with the realities here.
Our research has also paralleled the previous posts you mention from ELRS devs which explored power output, but we have taken it a step further and tested in the field, along with precise tools.

**Comment by SunjunKim** [comment]
> Our research has also paralleled the previous posts you mention from ELRS devs which explored power output, but we have taken it a step further and tested in the field, along with precise tools.

Field experience and tool-based tests are equally important.
We, ELRS team, have tried to be scientific as much as possible so far. This is to avoid making a wrong statement based on subjective *feelings*, but supported by objective cold facts.

If you could provide your measurement results along with the field testing results with some data, it'd be best to start a constructive discussion.

**Comment by Gorilla-Link** [comment]
Screenshot (1648)

Screenshot (1649)

Here are two field tests for penetration over 200 yards away and through a three-story college library building. We use the telemetry values to measure performance. Specifically the antenna switching (diversity antenna) and the RSNR. Hopefully this satisfies some concern for maximum power being sustainable for long durations.

Next we have measurement of output power. From this we see it is ~1.8 Watts which is the same value measured from the Radiomaster 2 Watt module, therefore it should be just as acceptable equipment and useful for dynamic power situations.

Screenshot (1675)

Screenshot (1676)

Hope this is helpful to anyone interested in testing their gear. It would be cool if all the other Vendors could also show some data from their testing..

For distance, we know that 100km is achievable with 2.4GHz ;)  so the 900 MHz system will be if anything, MORE robust.

**Comment by JyeSmith** [comment]
@Gorilla-Link could you share some measurements on all power settings e.g. 10 through 1000mW.  And for the same for the receiver, 10 - 100mW.

**Comment by JyeSmith** [comment]
Regardless of the output increase, there are still the concerns of using the sx1276 and rf amp outside of the recommended specs.  Can you imagine the fall out if these are used and hardware starts to fail.

File: src/include/target/Frsky_TX_R9M.h (hunk 1 above)
Comment (Gorilla-Link) [reply src/include/target/Frsky_TX_R9M.h:49]: 2900, aka 2.9 volts is where the peak dBm was achieved with our tests across many units. But we also saw no adverse effects when running at 3.3 volts, as there is overcurrent protection. As long as this is 2900 that should be fine, as 3.3 isn't really necessary either way.

**Comment by Gorilla-Link** [comment]
Would you prefer to change the name, to "experiments" like similar pull requests that exist.

Maybe there can be a separate branch? We understand your concerns making it part of the main.

Or what about the "unlock higher power" option box? This was available in the past.

Thanks for this consideration, and hopefully shouldn't be too much extra time.

**Comment by brandonrc** [comment]
@Gorilla-Link ExpressLRS has went from a very DIY firmware to something that professionals can trust while on a job. We spend a lot of time, unpaid time, testing and vetting new hardware. There is a process for getting new hardware added and approved by the dev team. 

Address 2W Power:
I think the dev team has always been against throwing more power at TX's just for the sake of it. 

> Would you prefer to change the name, to "experiments" like similar pull requests that exist.

`Experiments` pull requests should be used for PRs that the project plans to merge into main branch. 

> Maybe there can be a separate branch? We understand your concerns making it part of the main.

Seperate branch on the ExpressLRS repository still has to meet @JyeSmith @SunjunKim and the rest of the Devs approval. 

> Or what about the "unlock higher power" option box? This was available in the past.

`Unlock higher power` option box would lead users to believe that is a safe option to use. 

My opinion about power:
Higher power TX's create a false sense of security. Users believe they are going to get more range but the irony is if your TX becomes unstable then you will get less range or fail safes.
- More power is not always better
- Only use the power you need
- Safety

Also, why are we even talking about 2W on 900mhz? I believe there was a pilot who flew 100km's on 2.4ghz...  what is the use case of 2W with 900MHz? 

Lastly, you have the freedom to fork ExpressLRS and call it RecklessLRS. Add 3W mode if that makes you happy. You will have to maintain that and deal with the users who have issues. We just do not want to let a user easily use hardware outside of safety thresholds.

**Pull request closed without merging**